### PR TITLE
Deactivate plugin versions below 4.1 on WordPress 5.0

### PR DESCRIPTION
**Pull request by mcsf**
~**DO NOT MERGE.** This is meant to be a patch applied directly to previous releases. Read on.~
**Edit:** Patches released tagged. See https://github.com/WordPress/gutenberg/pull/11375#issuecomment-435220976

Ignore the diff against `master`. [See actual diff against v4.0.0.](https://github.com/WordPress/gutenberg/compare/v4.0.0...fix/deactivate-if-5-0)

## Description

Core ticket [45123](https://core.trac.wordpress.org/ticket/45123) took care of systematically deactivating Gutenberg upon upgrading to WordPress 5.0, both as a general precaution and due to the fact that core WordPress becomes the default provider of Gutenberg functionality.

By "general precaution" it is understood that, although core and plugin are expected to be compatible starting with Gutenberg 4.1, they naturally share a large surface and we want to avoid any chance of failure during a core upgrade process.

However, versions of the plugin prior to 4.1 are fundamentally incompatible with WordPress 5.0, to the point where [core's procedure of deactivating](https://core.trac.wordpress.org/browser/branches/5.0/src/wp-admin/includes/upgrade.php?rev=43765#L1802) the plugin happens too late to guarantee isolation, [potentially causing a site to WSOD](https://core.trac.wordpress.org/ticket/45157).

Thus, this PR proposes a patch for Gutenberg versions 3.6, 3.7, 3.8, 3.9 and 4.0 that deactivates itself if any version of WordPress above 4.9.8 is detected. It can do this before execution of any other code in the plugin, thus ensuring no Gutenberg code taints core's upgrade routine. With the help of @pento, patch releases will be pushed as auto-updates to existing installations.

The commit list shows an initial version that showed a deactivation notice, but in practice I found this to be useless.

## How has this been tested?
- Build the plugin from this patch (`bin/build-plugin-zip.sh`).
- Install and activate it on a WP 4.9.8 site.
- Confirm that the plugin stays activated and that the Gutenberg editor is present.
- Using the [Beta Tester Plugin](https://wordpress.org/plugins/wordpress-beta-tester/) or any comparable means, upgrade to a WP 5.0 nightly or beta.
- Ensure that the process goes smoothly and that it deactivates the Gutenberg plugin.

- Repeat the steps using a *non-patched* version of 4.0 or lower and confirm that the upgrade process is negatively affected by Gutenberg.

### CODE REVIEW

**Review by tofumatt — APPROVED** [review]
Wait, so this is patching previous versions of released Gutenberg? I guess it's just for releases in the plugin repo and not node packages, etc., right?

The code here seems fine (though I'll point out I haven't tested it myself), but I'm unclear on what's being updated and how.

Conceptually though seems good. 🤷‍♂️

**Comment by tofumatt** [comment]
(If you really need me to test the whole process I could do it, let me know, I'm just being lazy and triaging stuff and I don't quite understand the steps, sorry.)

**Review by jorgefilipecosta — APPROVED** [review]
I followed the test instructions testing the update on a zip build from this branch and Gutenberg deactivated and the update of WordPress was successful. Using the zip file from 4.0 rc available on GitHub the update of WordPress went wrong and I started getting error 500. So everything went as expected.

File: gutenberg.php [hunk 1]
```diff
@@ -3,12 +3,20 @@
...
  * Description: Printing since 1440. This is the development plugin for the new block editor in core.
- * Version: 4.0.0-rc.1
+ * Version: 4.0.0
  * Author: Gutenberg Team
  *
  * @package gutenberg
  */
 
+// Get unmodified $wp_version.
+include ABSPATH . WPINC . '/version.php';
```
[the remark this answers is not in the document]
Comment (mcsf) [reply gutenberg.php]: I had the same reflex about `include_once`, but for that particular file `include` may be best for two reasons:
- `version.php` should only contain variable assignments, so no errors from including multiple times should ever happen
- by using `include_once` it's theoretically possible that the inclusion gets skipped (due to prior inclusion) and that the `wp_version` variable has been overridden elsewhere. Not that I feel too paranoid about this.

**Comment by mcsf** [comment]
> Wait, so this is patching previous versions of released Gutenberg? I guess it's just for releases in the plugin repo and not node packages, etc., right?

Exactly, this is just a small patch for `gutenberg.php` that should be applied on top of each tag for `3.6` through `4.0` (as `3.6.3`, `3.7.1`, etc.), then tagged in SVN.

**Comment by mcsf** [comment]
Tagged:

- https://github.com/WordPress/gutenberg/releases/tag/v4.0.1
- https://github.com/WordPress/gutenberg/releases/tag/v3.9.1
- https://github.com/WordPress/gutenberg/releases/tag/v3.8.1
- https://github.com/WordPress/gutenberg/releases/tag/v3.7.1
- https://github.com/WordPress/gutenberg/releases/tag/v3.6.3

**Pull request closed without merging**